### DISCUSSION TITLE
Plot ensemble Kx-arrays as a top oriented model grid for each layer

### DISCUSSION THREAD

**Discussion by d-strom** [pypest/pyemu#477] [Q&A] [answered]
Hi,

I would like to plot my generated ensembles after a PEST-IES run in a way similar to the example in the GMDSI-tutorial example with Monte Carlo (see fig. below).

![image](https://github.com/pypest/pyemu/assets/74509984/d00c9d51-e6a5-44c3-9e1f-41c0e312b6c8)

Can anyone help me or point me in the direction of some classes/functions that would let me do this? Preferably in a simple way.

The reason for it is to have some intuitive interpretation of K-values and to make a screening of which arrays match my conceptual description of the studied magazine.

I'm working on a model where I'm struggling to understand why  PEST-IES isn't able to give me a better match with my observation-data since everything indicates that it should be. 

Sincerely
David

**Comment by javgs-bd** [comment]
Hi @DStrom1987. The way we go about this is by coding a loop where (loosely): 

i) read one parameter vector (one realization) from the resulting par.csv file.
ii) write a PEST par file

From there:

option a.- 
iii) pass this par file to the series of PEST templates cited in the PST file using TEMPCHEK
iv) Run the parameter preprocessing routine (for us is PLPROC)

option b.-
iii) use the PAR file with PARREP to generate a new PST file.
iv) run the model command section in the PST using NOPTMAX 0

Then
v) load the model with flopy and plot the K field

option a.- will require a temp folder for the dummy model to generate input files and load to flopy, while option b requires the temp folders for the manager and agents of the inversion loop.

There are a number of options for reading/writting files, folder creation and running processes within a python script. We tend to default for numpy genfromtext and the subprocess library. There are also examples using pandas and shutil.

Hope it helps
Javier

**Reply by rhugman** [reply]
Or add all your model cell properties as observations. OE can get a wee bit large, but usually manageable for testing purposes at least.

**Reply by d-strom** [reply]
Alright, thanks! I'll try to give this a go even though my programming could definitely need some polishing.. :)

**Reply by d-strom** [reply]
@rhugman 

Does the addition of properties as observations require the use of PstFrom? I have another topic open regarding PstFrom-bits, as I am employing a somewhat unorthodox workflow to generate .pst files externally, in conjunction with the pyemu suite.

I am uncertain about how to incorporate additional observations beyond those covered by the observation package in Modflow. I have watched webinars hosted by both you and Katie/Mike, where you discuss the inclusion of virtually anything in the observation ensemble, ranging from model convergence to the flow budget, and so on.

I haven't seen any practical demonstrations of this in any of the gmdsi-tutorial notebooks, at least. Are there any examples that illustrate how this could be accomplished? Such guidance would be greatly appreciated! :)

Sincerely

**Comment by wkitlasten** [comment]
I don't think you need PstFrom to add properties as obs. It should be as
easy as an ins file corresponding to the property file (plus addition of
those in the pst control file). A 2x2 array of parameters (e.g., K) might
be:

pif *
!obs00! !obs01!
!obs10! !obs11!

(Forgive me, it has been a long time since I wrote my own instruction file!)


"Perfect spheres are pointless."

**Reply by jtwhite79** [reply, marked as answer]
and with that instruction file (assume its named "my_obs.txt.ins" with a corresponding output file named "my_obs.txt"), you can add those model outputs to the control file like this:

```
pst = pyemu.Pst("pest.pst")
pst.add_observations("my_obs.txt.ins","my_obs.txt")
pst.write("pest_new.pst")

```
note these observations will be given an `obsval` that is the values found in "my_obs.txt" and a generic weight of 1.0

**Reply by d-strom** [reply]
Thanks! I'll give it a go.

**Comment by d-strom** [comment]
> and with that instruction file (assume its named "my_obs.txt.ins" with a corresponding output file named "my_obs.txt"), you can add those model outputs to the control file like this:
> 
> ```
> pst = pyemu.Pst("pest.pst")
> pst.add_observations("my_obs.txt.ins","my_obs.txt")
> pst.write("pest_new.pst")
> ```
> 
> note these observations will be given an `obsval` that is the values found in "my_obs.txt" and a generic weight of 1.0
